File: java/io/github/anaruto/libbbf/ArchiveBuilder.java
package io.github.anaruto.libbbf;

public class ArchiveBuilder implements AutoCloseable {
    static {
        System.loadLibrary("bbf");
    }

    private long handle;

    public ArchiveBuilder(String outputPath, ArchiveFormat format) {
        this.handle = createNative(outputPath, format.ordinal());
        if (this.handle == 0) {
            throw new RuntimeException("Failed to create ArchiveBuilder for " + outputPath);
        }
    }

    @Override
    public void close() {
        if (this.handle != 0) {
            closeNative(this.handle);
            this.handle = 0;
        }
    }

    public boolean addPage(String filePath, String nameInArchive) {
        if (this.handle == 0) {
            throw new IllegalStateException("Builder is closed or finalized");
        }
        return addPage(this.handle, filePath, nameInArchive);
    }

    public void finalizeBuilder() {
        if (this.handle == 0) {
            throw new IllegalStateException("Builder is closed or finalized");
        }
        boolean success = finalizeNative(this.handle);
        this.handle = 0; // Consumed by finalizeNative
        if (!success) {
            throw new RuntimeException("Failed to finalize archive container");
        }
    }

    private static native long createNative(String outputPath, int format);
    private static native void closeNative(long handle);
    private static native boolean addPage(long handle, String filePath, String nameInArchive);
    private static native boolean finalizeNative(long handle);
}

File: kotlin/io/github/anaruto/libbbf/ArchiveBuilder.kt
package io.github.anaruto.libbbf

class ArchiveBuilder(outputPath: String, format: ArchiveFormat) : AutoCloseable {
    private var handle: Long = 0L

    init {
        handle = createNative(outputPath, format.ordinal)
        if (handle == 0L) {
            throw RuntimeException("Failed to create ArchiveBuilder for $outputPath")
        }
    }

    override fun close() {
        if (handle != 0L) {
            closeNative(handle)
            handle = 0L
        }
    }

    fun addPage(filePath: String, nameInArchive: String): Boolean {
        check(handle != 0L) { "Builder is closed or finalized" }
        return addPage(handle, filePath, nameInArchive)
    }

    fun finalizeBuilder() {
        check(handle != 0L) { "Builder is closed or finalized" }
        val success = finalizeNative(handle)
        handle = 0L // consumed by finalizeNative
        if (!success) {
            throw RuntimeException("Failed to finalize archive container")
        }
    }

    companion object {
        init {
            System.loadLibrary("bbf")
        }

        @JvmStatic
        private external fun createNative(outputPath: String, format: Int): Long

        @JvmStatic
        private external fun closeNative(handle: Long)

        @JvmStatic
        private external fun addPage(handle: Long, filePath: String, nameInArchive: String): Boolean

        @JvmStatic
        private external fun finalizeNative(handle: Long): Boolean
    }
}
